### ISSUE TITLE
Enable optional prefixes or suffixes for some or all generated classes

### ISSUE THREAD

**Issue by ddobrev** [semlanik/qtprotobuf#195]
**Problem description**
Generated classes, being mostly serializers, are nothing more than bags of data. To make them more object-oriented, one might want to add custom code. At present this is only possible through inheritance. This means new names are required for the subclasses and since it's these subclasses that will be used directly, they won't have the good names which map exactly to protobuf messages. This is why it would help to be able to rename generated classes so that the best names are available to users. It's important to be able to only rename some classes because not all of them would need extra functionality and therefore subclasses.

**Suggestion/solution**
Some way to add prefixes or suffixes to some or all generated classes so that the real names remain available.

**Alternatives**
Using substandard names for actually used subclasses while the good names remain unused.

**Additional context**
Not connected to renaming itself but to the origin of the idea is that deserializers should be somehow able to fill subclasses as well. If they can only work with the generated type, this means that a subclass must be created by copying data from the protobuf which is error-prone and might affect performance.

**Comment by ddobrev** [comment]
I think any modified name should also apply to the name of the file too - when a file's generated per class/message.

**Comment by semlanik** [comment]
@ddobrev to keep you posted, I've tried implementing this feature in an ogre style(burst into the room and smashed all the dishes). So there are several issues that I need to solve, but the main issue is in the implementation of the class namespaces and the nested types. Currently, namespace sequences are resolved using the full name of a message. So I need to identify how to inject custom prefixes/suffixes into the namespaces since the full message name is specified as a single dot-separated string. This doesn't give me an explicit way to identify the message name among namespace related stuff.

**Comment by semlanik** [comment]
Now I'm wondering about profits that you may have when using prefixes/suffixes and couldn't they be replaced by the highest-level namespace?

So if you want to inherit the protobuf message in your project you may do this:

``` cpp
class somePackage::MessageName : public **generatedMessagesNamespace**::somePackage::MessageName
{
};
```

**Comment by gmabey** [comment]
Personally I do not feel that the QtProtobuf data containers fit into an inheritance-centric design.  After all, they are *containers* not *actors*.  That is, in "extending" a class, you are adding more features (behavior and data) that are a superset of the base class.  That doesn't jive with the protobuf message, because you _should_ _not_ be extending the behavior or data of the message in this way -- instead, you should go back and augment your `.proto` file.

Instead, the generated classes should be a data member of a class that (itself) can be extended using C++ inheritance.

My 2 cents ...

**Comment by semlanik** [comment]
@gmabey overall I agree on this point since all manipulations with internals should remain encapsulated into generated code. I'm only guessing, but perhaps @ddobrev is trying to implement the Facade pattern for the existing code, so it makes sense to allow doing this.

**Comment by ddobrev** [comment]
@gmabey generated objects are too dumb, sometimes additional functionality is required for real OOP. This is why there has to be a way to add custom code.
@semlanik in general I don't mind but there's a problem. Some objects would need no extensions which means they'd be used directly. This means their name-spaces would be required thus creating a conflict with the custom class unless it's always used with full qualification which is inconvenient.
I do see you've hit an obstacle, though. I'll see if I can think of a solution to help you.

**Comment by semlanik** [comment]
@ddobrev in general with prefixes and suffixes you will have to use the addition to the class name in any case. I don't see the difference between the fully-qualified class name generated::ClassName and the suffixed/prefixed name ClassNameGenerated.

I found another major problem - resolving the cross-proto file dependencies. So if you have a.proto that contains message A and it's generated as <custom_suffix>A and then you have b.proto that includes a.proto to use A as a field of the message B, this will cause an issue, that generator may not resolve. Of course, this will not happen if the generating for the a.proto and b.proto will happen with the same parameters or using the same call of prootoc, but it also opens a room for unwanted issues in the project.

In conclusion, I could only offer the generating of the extra namespaces in the scope of the single protoc call, that covers all messages in proposed .proto files and the result of using any includes will be unpredictable.

**Comment by ddobrev** [comment]
I understand, @semlanik. Please do it any way you can, I'm thankful for your work. Besides, I'm getting impatient to try this so let's go ahead with 0.6.0 as soon as possible. 🙂

**Comment by ddobrev** [comment]
@semlanik so pleasing to see your work. 🙂 Can this new feature only work for certain classes? In most cases the overwhelming majority of generated classes are directly usable and just a few would require tweaks and therefore inheritance. This is what I meant by "some" in the title.

**Comment by semlanik** [comment]
@ddobrev , here I need some input from you, how you would like to mark such classes? I could suggest the option mechanism. But this is something completely new to implement(First I need to check how it works). And as I already mentioned this may cause an issue when such a message is used as another message field.

**Comment by ddobrev** [comment]
I'm not sure I can suggest given that I'm no expert in CMake. Perhaps some kind of dictionary where the name of the class is the key and the name-space is the value. Or, if possible, make that `EXTRA_NAMESPACE` of yours take a list of classes as an extra parameter, assuming all if the list is empty. Is any of this possible in `CMake`?

**Comment by semlanik** [comment]
It's not related to the CMake but to the generator logic. To keep generated classes consistent these "extras" have to be specified on a .proto file level, otherwise, there will be a naming mismatch between generated messages. Also, looks like even the current approach breaks nesting and maps as prefixes/suffixes did.

**Comment by ddobrev** [comment]
@semlanik I'm really sorry for making you do this. I had no idea you'd have this amount of trouble. Please just go ahead with 0.6.0 - I'll get over one or two suboptimal names.

**Comment by semlanik** [comment]
@ddobrev , nothing to sorry about :) It's only way I excuse myself for the slow feature development.
